src/uct/ugni/smsg/ugni_smsg_iface.c

```diff
@@ -0,0 +1,432 @@
+    }
+
+    GNI_SmsgRelease(ep->super.ep);
+  }
+}
+
+static void progress_remote_cq(uct_ugni_smsg_iface_t *iface){
+  gni_return_t ugni_rc;
+  gni_cq_entry_t event_data;
+  uct_ugni_ep_t tl_ep;
+  uct_ugni_ep_t *ugni_ep;
+  uct_ugni_smsg_ep_t *ep;
+
+  ugni_rc = GNI_CqGetEvent(iface->remote_cq, &event_data);
+
+  if(GNI_RC_NOT_DONE == ugni_rc){
+    return;
+  }
+
+  if (GNI_RC_SUCCESS != ugni_rc || GNI_CQ_OVERRUN(event_data)) {
+    ucs_error("GNI_CqGetEvent falied. Error status %s %d ",
+              gni_err_str[ugni_rc], ugni_rc);
+    return;
+  }
+
+  tl_ep.hash_key = GNI_CQ_GET_INST_ID(event_data);
+  ugni_ep = sglib_hashed_uct_ugni_ep_t_find_member(iface->super.eps, &tl_ep);
+  ep = ucs_derived_of(ugni_ep, uct_ugni_smsg_ep_t);
+
+  process_mbox(iface, ep);
+}
+
+UCS_CLASS_DEFINE_DELETE_FUNC(uct_ugni_smsg_iface_t, uct_iface_t);
+
+static void uct_ugni_smsg_progress(void *arg)
+{
+  uct_ugni_smsg_iface_t *iface = (uct_ugni_smsg_iface_t *)arg;
+
+  progress_local_cq(iface);
+  progress_remote_cq(iface);
+
+  /* have a go a processing the pending queue */
+  ucs_arbiter_dispatch(&iface->super.arbiter, 1, uct_ugni_ep_process_pending, NULL);
+}
+
+static void uct_ugni_smsg_iface_release_am_desc(uct_iface_t *tl_iface, void *desc)
+{
+  uct_ugni_smsg_desc_t *ugni_desc = ((uct_ugni_smsg_desc_t *)desc)-1;
+  ucs_mpool_put(ugni_desc);
+}
+
+static ucs_status_t uct_ugni_smsg_query_tl_resources(uct_pd_h pd,
+                                                     uct_tl_resource_desc_t **resource_p,
+                                                     unsigned *num_resources_p)
+{
+  return uct_ugni_query_tl_resources(pd, UCT_UGNI_SMSG_TL_NAME,
+                                     40000, /* nano sec */
+                                     pow(1024,2), /* bytes */
+                                     resource_p, num_resources_p);
+}
+
+static ucs_status_t uct_ugni_smsg_iface_query(uct_iface_h tl_iface, uct_iface_attr_t *iface_attr)
+{
+  uct_ugni_smsg_iface_t *iface = ucs_derived_of(tl_iface, uct_ugni_smsg_iface_t);
+
+  memset(iface_attr, 0, sizeof(uct_iface_attr_t));
+  iface_attr->cap.am.max_short       = iface->config.smsg_seg_size-sizeof(uint64_t);
+  iface_attr->cap.am.max_bcopy       = iface->config.smsg_seg_size;
+  iface_attr->iface_addr_len         = sizeof(uct_sockaddr_ugni_t);
+  iface_attr->ep_addr_len            = sizeof(uct_sockaddr_smsg_ugni_t);
+  iface_attr->cap.flags              = UCT_IFACE_FLAG_AM_SHORT |
+                                       UCT_IFACE_FLAG_AM_BCOPY |
+                                       UCT_IFACE_FLAG_CONNECT_TO_EP |
+                                       UCT_IFACE_FLAG_AM_THREAD_SINGLE;
+  return UCS_OK;
+}
+
+
+static UCS_CLASS_CLEANUP_FUNC(uct_ugni_smsg_iface_t)
+{
+  uct_worker_progress_unregister(self->super.super.worker,
+                                 uct_ugni_smsg_progress, self);
+  if (!self->super.activated) {
+    return;
+  }
+  ucs_mpool_put(self->user_desc);
```

Comment:
Most of that is in the EPs rather than iface, since the EPs are the unit that knows what the mbox looks like.